        /**
         * Will throw an exception if the wait was terminated with an error. If it was canceled, this method will
         * do nothing.
         */
        public void throwExceptionIfNeeded() {
            if (resultReceived && errorCode != null) {
                throw new AppException(ErrorCode.UNKNOWN,
                        String.format(Locale.US, "Internal error (%d): %s", errorCode, errorMessage));
            }
        }

